odoo/addons/base/models/res_users.py

```diff
@@ -199,6 +199,84 @@ def _search(self, args, offset=0, limit=None, order=None, count=False, access_ri
+                selected_group.category_id != group.category_id and
+                selected_group not in group.implied_ids
+            )
+            for selected_group in self
+        }
+        warnings = []
+        for selected_group, other_groups in inherited_groups.items():
+            if not other_groups:
+                continue
+            # From the groups user has, find the ones with selected group's category
+            selected_category_groups = self.filtered(lambda group: group.category_id == selected_group.category_id)
+            # Get all the implied groups of a selected group
+            selected_category_groups |= selected_category_groups._get_implied_groups()
+            # Get the minimum group needed in selected group's category (For example, someone changes
+            # Sales: Admin to Sales: User, but Field Service is already set to Admin, so here in the
+            # 'Sales' category, we will at the minimum need Admin group)
+            minimum_required_groups = other_groups.implied_ids.filtered(
+                lambda group:
+                group.category_id == selected_group.category_id and
+                group not in selected_category_groups
+            )
+            if not minimum_required_groups:
+                continue
+            all_selected_category_groups = self.search([('category_id', '=', selected_group.category_id.id)])
+            # Groups that have higher access level than minimum_required_groups
+            possible_groups = all_selected_category_groups - minimum_required_groups._get_implied_groups()
+            warnings.append(_(
+                "Since %s is a/an %s %s, you can set %s right only to %s"
+                ) % (
+                    user.name,
+                    other_groups[0].category_id.name,
+                    other_groups[0].name,
+                    selected_group.category_id.name,
+                    _(" or ").join(possible_groups.mapped('name'))
+                )
+            )
+        return warnings
+
+    def _prepare_warnings_for_removed_groups(self, user, categories):
+        warnings = []
+        for category in categories:
+            # Find the group that enforces that at-least 'minimum_required_groups' in empty categories
+            enforcing_groups = self.filtered(lambda grp: category in grp.implied_ids.category_id)
+            if not enforcing_groups:
+                continue
+            # Get the minimum group needed for emptied category
+            minimum_required_groups = enforcing_groups.implied_ids.filtered(lambda grp: grp.category_id == category)
+            warnings.append(_(
+                "Since %s is a/an %s %s, they will at the minimum have the %s: %s access too"
+                ) % (
+                    user.name,
+                    enforcing_groups[0].category_id.name,
+                    enforcing_groups[0].name,
+                    category.name,
+                    minimum_required_groups[0].name
```

Comment:
is it technically possible to have `minimum_required_groups` empty here ? (if yes, `minimum_required_groups[0]` will crash in that case)